index.js

```diff
@@ -1,175 +1,249 @@
 
-Object.defineProperty(LiveReloadPlugin.prototype, 'isRunning', {
-  get: function() { return !!this.server; }
-});
+    compiler.hooks.compilation.tap('LiveReloadPlugin', this._applyCompilation.bind(this));
+    compiler.hooks.watchRun.tapAsync('LiveReloadPlugin', this._start.bind(this));
+    compiler.hooks.afterEmit.tap('LiveReloadPlugin', this._afterEmit.bind(this))
+    compiler.hooks.failed.tap('LiveReloadPlugin', this._failed.bind(this));
+  }
 
-function generateHashCode(str) {
-  const hash = crypto.createHash('sha256');
-  hash.update(str);
-  return hash.digest('hex');
-}
+  /**
+   * @param a1
+   * @param a2
+   * @returns {boolean|*}
+   */
+  static arraysEqual(a1, a2) {
+    return a1.length === a2.length && a1.every((v,i) => v === a2[i])
+  }
 
-function fileIgnoredOrNotEmitted(data) {
-  if (Array.isArray(this.ignore)) {
-    return !anymatch(this.ignore, data[0]) && data[1].emitted;
+  /**
+   * @param str
+   * @returns {string}
+   */
+  static generateHashCode(str) {
+    const hash = crypto.createHash('sha256');
+    hash.update(str);
+    return hash.digest('hex');
   }
-  return !data[0].match(this.ignore) && data[1].emitted;
-}
 
-function fileHashDoesntMatches(data) {
-  if (!this.useSourceHash)
-    return true;
+  /**
+   * @param compilation
+   * @returns {never}
+   * @private
+   */
+  _applyCompilation(compilation) {
+    if (this.isWebpack4) {
+      return compilation.mainTemplate.hooks.startup.tap('LiveReloadPlugin', this._scriptTag.bind(this));
+    }
 
-  const sourceHash = generateHashCode(data[1].source());
-  if (
-      this.sourceHashs.hasOwnProperty(data[0])
-      && this.sourceHashs[data[0]] === sourceHash
-  ) {
-    return false;
+    this.webpack.javascript.JavascriptModulesPlugin.getCompilationHooks(compilation).renderRequire.tap('LiveReloadPlugin', this._scriptTag.bind(this));
   }
 
-  this.sourceHashs[data[0]] = sourceHash;
-  return true;
-};
+  /**
+   * @param watching
+   * @param cb
+   * @private
+   */
+  _start(watching, cb) {
+    const quiet = this.options.quiet;
+    if (servers[this.options.port]) {
+      this.server = servers[this.options.port];
+      cb();
+    } else {
+      const listen = () => {
+        const options = this.options;
+        delete options.delay;
+        this.server = servers[this.options.port] = lr(options);
+
+        this.server.errorListener = (err) => {
+          console.error('Live Reload disabled: ' + err.message);
+          if (err.code !== 'EADDRINUSE') {
+            console.error(err.stack);
+          }
+          cb();
+        };
+
+        this.server.listen(this.options.port, (err) => {
+          if (!err && !quiet) {
+            console.log('Live Reload listening on port ' + this.options.port + '\n');
+          }
+          cb();
+        });
+      };
+
+      if(this.options.port === 0) {
+        portfinder.basePort = this.defaultPort;
+        portfinder.getPort((err, port) => {
+          if (err) {
+            throw err;
```

Comment:
Not your change, but I noticed this while reviewing. For playing nice as a plugin, this error should be sent up through the callback instead of throwing.

```suggestion
            return cb(err);
```